// @mastra/server/dist/server/handlers/mcp.cjs
'use strict';

var chunkWJ6DNHLW_cjs = require('../../chunk-WJ6DNHLW.cjs');



Object.defineProperty(exports, "EXECUTE_MCP_SERVER_TOOL_ROUTE", {
  enumerable: true,
  get: function () { return chunkWJ6DNHLW_cjs.EXECUTE_MCP_SERVER_TOOL_ROUTE; }
});
Object.defineProperty(exports, "GET_MCP_SERVER_DETAIL_ROUTE", {
  enumerable: true,
  get: function () { return chunkWJ6DNHLW_cjs.GET_MCP_SERVER_DETAIL_ROUTE; }
});
Object.defineProperty(exports, "GET_MCP_SERVER_TOOL_DETAIL_ROUTE", {
  enumerable: true,
  get: function () { return chunkWJ6DNHLW_cjs.GET_MCP_SERVER_TOOL_DETAIL_ROUTE; }
});
Object.defineProperty(exports, "LIST_MCP_SERVERS_ROUTE", {
  enumerable: true,
  get: function () { return chunkWJ6DNHLW_cjs.LIST_MCP_SERVERS_ROUTE; }
});
Object.defineProperty(exports, "LIST_MCP_SERVER_TOOLS_ROUTE", {
  enumerable: true,
  get: function () { return chunkWJ6DNHLW_cjs.LIST_MCP_SERVER_TOOLS_ROUTE; }
});
Object.defineProperty(exports, "MCP_HTTP_TRANSPORT_ROUTE", {
  enumerable: true,
  get: function () { return chunkWJ6DNHLW_cjs.MCP_HTTP_TRANSPORT_ROUTE; }
});
Object.defineProperty(exports, "MCP_SSE_MESSAGES_ROUTE", {
  enumerable: true,
  get: function () { return chunkWJ6DNHLW_cjs.MCP_SSE_MESSAGES_ROUTE; }
});
Object.defineProperty(exports, "MCP_SSE_TRANSPORT_ROUTE", {
  enumerable: true,
  get: function () { return chunkWJ6DNHLW_cjs.MCP_SSE_TRANSPORT_ROUTE; }
});
