// @mastra/deployer/dist/chunk-5TYVF4KJ.cjs
'use strict';

var path = require('path');
var module$1 = require('module');

// src/build/utils.ts
function detectRuntime() {
  if (process.versions?.bun) {
    return "bun";
  }
  return "node";
}
function isDependencyPartOfPackage(dep, packageName) {
  if (dep === packageName) {
    return true;
  }
  return dep.startsWith(`${packageName}/`);
}
function getPackageName(id) {
  const parts = id.split("/");
  if (id.startsWith("@")) {
    return parts.slice(0, 2).join("/");
  }
  return parts[0];
}
function getCompiledDepCachePath(rootPath, packageName) {
  return slash(path.join(rootPath, "node_modules", ".cache", packageName));
}
function slash(path) {
  const isExtendedLengthPath = path.startsWith("\\\\?\\");
  if (isExtendedLengthPath) {
    return path;
  }
  return path.replaceAll("\\", "/");
}
function rollupSafeName(name, rootDir) {
  const rel = path.relative(rootDir, name);
  let entry = slash(rel);
  entry = entry.replace(/^(\.\.\/)+/, "");
  entry = entry.replace(/^\/+/, "");
  entry = entry.replace(/^[A-Za-z]:\//, "");
  if (!entry) {
    entry = slash(path.basename(name));
  }
  return entry;
}
function normalizeStudioBase(studioBase) {
  if (studioBase.includes("..") || studioBase.includes("?") || studioBase.includes("#")) {
    throw new Error(`Invalid base path: "${studioBase}". Base path cannot contain '..', '?', or '#'`);
  }
  studioBase = studioBase.replace(/\/+/g, "/");
  if (studioBase === "/" || studioBase === "") {
    return "";
  }
  if (studioBase.endsWith("/")) {
    studioBase = studioBase.slice(0, -1);
  }
  if (!studioBase.startsWith("/")) {
    studioBase = `/${studioBase}`;
  }
  return studioBase;
}
function isBuiltinModule(specifier) {
  return module$1.builtinModules.includes(specifier) || specifier.startsWith("node:") || module$1.builtinModules.includes(specifier.replace(/^node:/, ""));
}

exports.detectRuntime = detectRuntime;
exports.getCompiledDepCachePath = getCompiledDepCachePath;
exports.getPackageName = getPackageName;
exports.isBuiltinModule = isBuiltinModule;
exports.isDependencyPartOfPackage = isDependencyPartOfPackage;
exports.normalizeStudioBase = normalizeStudioBase;
exports.rollupSafeName = rollupSafeName;
exports.slash = slash;
