// @mastra/server/dist/chunk-FLLJIPD4.cjs
'use strict';

var chunkAVRO3FIN_cjs = require('./chunk-AVRO3FIN.cjs');
var chunkSUKFO7UM_cjs = require('./chunk-SUKFO7UM.cjs');
var fs = require('fs');
var zod = require('zod');

var mastraPackageSchema = zod.z.object({
  name: zod.z.string(),
  version: zod.z.string()
});
var systemPackagesResponseSchema = zod.z.object({
  packages: zod.z.array(mastraPackageSchema)
});

// src/server/handlers/system.ts
var GET_SYSTEM_PACKAGES_ROUTE = chunkAVRO3FIN_cjs.createRoute({
  method: "GET",
  path: "/api/system/packages",
  responseType: "json",
  responseSchema: systemPackagesResponseSchema,
  summary: "Get installed Mastra packages",
  description: "Returns a list of all installed Mastra packages and their versions from the project",
  tags: ["System"],
  handler: async () => {
    try {
      const packagesFilePath = process.env.MASTRA_PACKAGES_FILE;
      let packages = [];
      if (packagesFilePath) {
        try {
          const fileContent = fs.readFileSync(packagesFilePath, "utf-8");
          packages = JSON.parse(fileContent);
        } catch {
          packages = [];
        }
      }
      return { packages };
    } catch (error) {
      return chunkSUKFO7UM_cjs.handleError(error, "Error getting system packages");
    }
  }
});

exports.GET_SYSTEM_PACKAGES_ROUTE = GET_SYSTEM_PACKAGES_ROUTE;
