// @mastra/server/dist/chunk-XCEQ4GQW.cjs
'use strict';

var chunkMCYD5LW7_cjs = require('./chunk-MCYD5LW7.cjs');
var chunkAVRO3FIN_cjs = require('./chunk-AVRO3FIN.cjs');
var chunkSUKFO7UM_cjs = require('./chunk-SUKFO7UM.cjs');
var chunk64ITUOXI_cjs = require('./chunk-64ITUOXI.cjs');
var z = require('zod');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var z__default = /*#__PURE__*/_interopDefault(z);

var storedAgentIdPathParams = z__default.default.object({
  storedAgentId: z__default.default.string().describe("Unique identifier for the stored agent")
});
var storageOrderBySchema = z__default.default.object({
  field: z__default.default.enum(["createdAt", "updatedAt"]).optional(),
  direction: z__default.default.enum(["ASC", "DESC"]).optional()
});
var listStoredAgentsQuerySchema = chunkMCYD5LW7_cjs.createPagePaginationSchema(100).extend({
  orderBy: storageOrderBySchema.optional()
});
var scorerConfigSchema = z__default.default.object({
  sampling: z__default.default.object({
    type: z__default.default.enum(["ratio", "count"]),
    rate: z__default.default.number().optional(),
    count: z__default.default.number().optional()
  }).optional()
});
var storedAgentBaseSchema = z__default.default.object({
  name: z__default.default.string().describe("Name of the agent"),
  description: z__default.default.string().optional().describe("Description of the agent"),
  instructions: z__default.default.string().describe("System instructions for the agent"),
  model: z__default.default.record(z__default.default.string(), z__default.default.unknown()).describe("Model configuration (provider, name, etc.)"),
  tools: z__default.default.array(z__default.default.string()).optional().describe("Array of tool keys to resolve from Mastra registry"),
  defaultOptions: z__default.default.record(z__default.default.string(), z__default.default.unknown()).optional().describe("Default options for generate/stream calls"),
  workflows: z__default.default.array(z__default.default.string()).optional().describe("Array of workflow keys to resolve from Mastra registry"),
  agents: z__default.default.array(z__default.default.string()).optional().describe("Array of agent keys to resolve from Mastra registry"),
  inputProcessors: z__default.default.array(z__default.default.record(z__default.default.string(), z__default.default.unknown())).optional().describe("Input processor configurations"),
  outputProcessors: z__default.default.array(z__default.default.record(z__default.default.string(), z__default.default.unknown())).optional().describe("Output processor configurations"),
  memory: z__default.default.string().optional().describe("Memory key to resolve from Mastra registry"),
  scorers: z__default.default.record(z__default.default.string(), scorerConfigSchema).optional().describe("Scorer keys with optional sampling config"),
  metadata: z__default.default.record(z__default.default.string(), z__default.default.unknown()).optional().describe("Additional metadata for the agent")
});
var createStoredAgentBodySchema = storedAgentBaseSchema.extend({
  id: z__default.default.string().describe("Unique identifier for the agent")
});
var updateStoredAgentBodySchema = storedAgentBaseSchema.partial();
var storedAgentSchema = storedAgentBaseSchema.extend({
  id: z__default.default.string(),
  createdAt: z__default.default.date(),
  updatedAt: z__default.default.date()
});
var listStoredAgentsResponseSchema = chunkMCYD5LW7_cjs.paginationInfoSchema.extend({
  agents: z__default.default.array(storedAgentSchema)
});
var getStoredAgentResponseSchema = storedAgentSchema;
var createStoredAgentResponseSchema = storedAgentSchema;
var updateStoredAgentResponseSchema = storedAgentSchema;
var deleteStoredAgentResponseSchema = z__default.default.object({
  success: z__default.default.boolean(),
  message: z__default.default.string()
});

// src/server/handlers/stored-agents.ts
var LIST_STORED_AGENTS_ROUTE = chunkAVRO3FIN_cjs.createRoute({
  method: "GET",
  path: "/api/stored/agents",
  responseType: "json",
  queryParamSchema: listStoredAgentsQuerySchema,
  responseSchema: listStoredAgentsResponseSchema,
  summary: "List stored agents",
  description: "Returns a paginated list of all agents stored in the database",
  tags: ["Stored Agents"],
  handler: async ({ mastra, page, perPage, orderBy }) => {
    try {
      const storage = mastra.getStorage();
      if (!storage) {
        throw new chunk64ITUOXI_cjs.HTTPException(500, { message: "Storage is not configured" });
      }
      const agentsStore = await storage.getStore("agents");
      if (!agentsStore) {
        throw new chunk64ITUOXI_cjs.HTTPException(500, { message: "Agents storage domain is not available" });
      }
      const result = await agentsStore.listAgents({
        page,
        perPage,
        orderBy
      });
      return result;
    } catch (error) {
      return chunkSUKFO7UM_cjs.handleError(error, "Error listing stored agents");
    }
  }
});
var GET_STORED_AGENT_ROUTE = chunkAVRO3FIN_cjs.createRoute({
  method: "GET",
  path: "/api/stored/agents/:storedAgentId",
  responseType: "json",
  pathParamSchema: storedAgentIdPathParams,
  responseSchema: getStoredAgentResponseSchema,
  summary: "Get stored agent by ID",
  description: "Returns a specific agent from storage by its unique identifier",
  tags: ["Stored Agents"],
  handler: async ({ mastra, storedAgentId }) => {
    try {
      const storage = mastra.getStorage();
      if (!storage) {
        throw new chunk64ITUOXI_cjs.HTTPException(500, { message: "Storage is not configured" });
      }
      const agentsStore = await storage.getStore("agents");
      if (!agentsStore) {
        throw new chunk64ITUOXI_cjs.HTTPException(500, { message: "Agents storage domain is not available" });
      }
      const agent = await agentsStore.getAgentById({ id: storedAgentId });
      if (!agent) {
        throw new chunk64ITUOXI_cjs.HTTPException(404, { message: `Stored agent with id ${storedAgentId} not found` });
      }
      return agent;
    } catch (error) {
      return chunkSUKFO7UM_cjs.handleError(error, "Error getting stored agent");
    }
  }
});
var CREATE_STORED_AGENT_ROUTE = chunkAVRO3FIN_cjs.createRoute({
  method: "POST",
  path: "/api/stored/agents",
  responseType: "json",
  bodySchema: createStoredAgentBodySchema,
  responseSchema: createStoredAgentResponseSchema,
  summary: "Create stored agent",
  description: "Creates a new agent in storage with the provided configuration",
  tags: ["Stored Agents"],
  handler: async ({
    mastra,
    id,
    name,
    description,
    instructions,
    model,
    tools,
    defaultOptions,
    workflows,
    agents,
    inputProcessors,
    outputProcessors,
    memory,
    scorers,
    metadata
  }) => {
    try {
      const storage = mastra.getStorage();
      if (!storage) {
        throw new chunk64ITUOXI_cjs.HTTPException(500, { message: "Storage is not configured" });
      }
      const agentsStore = await storage.getStore("agents");
      if (!agentsStore) {
        throw new chunk64ITUOXI_cjs.HTTPException(500, { message: "Agents storage domain is not available" });
      }
      const existing = await agentsStore.getAgentById({ id });
      if (existing) {
        throw new chunk64ITUOXI_cjs.HTTPException(409, { message: `Agent with id ${id} already exists` });
      }
      const toolsFromBody = Array.isArray(tools) ? tools : void 0;
      const agent = await agentsStore.createAgent({
        agent: {
          id,
          name,
          description,
          instructions,
          model,
          tools: toolsFromBody,
          defaultOptions,
          workflows,
          agents,
          inputProcessors,
          outputProcessors,
          memory,
          scorers,
          metadata
        }
      });
      return agent;
    } catch (error) {
      return chunkSUKFO7UM_cjs.handleError(error, "Error creating stored agent");
    }
  }
});
var UPDATE_STORED_AGENT_ROUTE = chunkAVRO3FIN_cjs.createRoute({
  method: "PATCH",
  path: "/api/stored/agents/:storedAgentId",
  responseType: "json",
  pathParamSchema: storedAgentIdPathParams,
  bodySchema: updateStoredAgentBodySchema,
  responseSchema: updateStoredAgentResponseSchema,
  summary: "Update stored agent",
  description: "Updates an existing agent in storage with the provided fields",
  tags: ["Stored Agents"],
  handler: async ({
    mastra,
    storedAgentId,
    name,
    description,
    instructions,
    model,
    tools,
    defaultOptions,
    workflows,
    agents,
    inputProcessors,
    outputProcessors,
    memory,
    scorers,
    metadata
  }) => {
    try {
      const storage = mastra.getStorage();
      if (!storage) {
        throw new chunk64ITUOXI_cjs.HTTPException(500, { message: "Storage is not configured" });
      }
      const agentsStore = await storage.getStore("agents");
      if (!agentsStore) {
        throw new chunk64ITUOXI_cjs.HTTPException(500, { message: "Agents storage domain is not available" });
      }
      const existing = await agentsStore.getAgentById({ id: storedAgentId });
      if (!existing) {
        throw new chunk64ITUOXI_cjs.HTTPException(404, { message: `Stored agent with id ${storedAgentId} not found` });
      }
      const toolsFromBody = Array.isArray(tools) ? tools : void 0;
      const agent = await agentsStore.updateAgent({
        id: storedAgentId,
        name,
        description,
        instructions,
        model,
        tools: toolsFromBody,
        defaultOptions,
        workflows,
        agents,
        inputProcessors,
        outputProcessors,
        memory,
        scorers,
        metadata
      });
      return agent;
    } catch (error) {
      return chunkSUKFO7UM_cjs.handleError(error, "Error updating stored agent");
    }
  }
});
var DELETE_STORED_AGENT_ROUTE = chunkAVRO3FIN_cjs.createRoute({
  method: "DELETE",
  path: "/api/stored/agents/:storedAgentId",
  responseType: "json",
  pathParamSchema: storedAgentIdPathParams,
  responseSchema: deleteStoredAgentResponseSchema,
  summary: "Delete stored agent",
  description: "Deletes an agent from storage by its unique identifier",
  tags: ["Stored Agents"],
  handler: async ({ mastra, storedAgentId }) => {
    try {
      const storage = mastra.getStorage();
      if (!storage) {
        throw new chunk64ITUOXI_cjs.HTTPException(500, { message: "Storage is not configured" });
      }
      const agentsStore = await storage.getStore("agents");
      if (!agentsStore) {
        throw new chunk64ITUOXI_cjs.HTTPException(500, { message: "Agents storage domain is not available" });
      }
      const existing = await agentsStore.getAgentById({ id: storedAgentId });
      if (!existing) {
        throw new chunk64ITUOXI_cjs.HTTPException(404, { message: `Stored agent with id ${storedAgentId} not found` });
      }
      await agentsStore.deleteAgent({ id: storedAgentId });
      return { success: true, message: `Agent ${storedAgentId} deleted successfully` };
    } catch (error) {
      return chunkSUKFO7UM_cjs.handleError(error, "Error deleting stored agent");
    }
  }
});

exports.CREATE_STORED_AGENT_ROUTE = CREATE_STORED_AGENT_ROUTE;
exports.DELETE_STORED_AGENT_ROUTE = DELETE_STORED_AGENT_ROUTE;
exports.GET_STORED_AGENT_ROUTE = GET_STORED_AGENT_ROUTE;
exports.LIST_STORED_AGENTS_ROUTE = LIST_STORED_AGENTS_ROUTE;
exports.UPDATE_STORED_AGENT_ROUTE = UPDATE_STORED_AGENT_ROUTE;
