// @mastra/server/dist/chunk-BRI4GUZ2.cjs
'use strict';

var chunkMCYD5LW7_cjs = require('./chunk-MCYD5LW7.cjs');
var z = require('zod');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var z__default = /*#__PURE__*/_interopDefault(z);

var agentIdPathParams = z__default.default.object({
  agentId: z__default.default.string().describe("Unique identifier for the agent")
});
var toolIdPathParams = z__default.default.object({
  toolId: z__default.default.string().describe("Unique identifier for the tool")
});
var agentToolPathParams = agentIdPathParams.extend({
  toolId: z__default.default.string().describe("Unique identifier for the tool")
});
var modelConfigIdPathParams = agentIdPathParams.extend({
  modelConfigId: z__default.default.string().describe("Unique identifier for the model configuration")
});
var serializedProcessorSchema = z__default.default.object({
  id: z__default.default.string(),
  name: z__default.default.string().optional()
});
var serializedToolSchema = z__default.default.object({
  id: z__default.default.string(),
  description: z__default.default.string().optional(),
  inputSchema: z__default.default.string().optional(),
  outputSchema: z__default.default.string().optional(),
  requireApproval: z__default.default.boolean().optional()
});
var serializedWorkflowSchema = z__default.default.object({
  name: z__default.default.string(),
  steps: z__default.default.record(
    z__default.default.string(),
    z__default.default.object({
      id: z__default.default.string(),
      description: z__default.default.string().optional()
    })
  ).optional()
});
var serializedAgentDefinitionSchema = z__default.default.object({
  id: z__default.default.string(),
  name: z__default.default.string()
});
var systemMessageSchema = z__default.default.union([
  z__default.default.string(),
  z__default.default.array(z__default.default.string()),
  z__default.default.any(),
  // CoreSystemMessage or SystemModelMessage
  z__default.default.array(z__default.default.any())
]);
var modelConfigSchema = z__default.default.object({
  model: z__default.default.object({
    modelId: z__default.default.string(),
    provider: z__default.default.string(),
    modelVersion: z__default.default.string()
  })
  // Additional fields from AgentModelManagerConfig can be added here
});
var serializedAgentSchema = z__default.default.object({
  name: z__default.default.string(),
  description: z__default.default.string().optional(),
  instructions: systemMessageSchema.optional(),
  tools: z__default.default.record(z__default.default.string(), serializedToolSchema),
  agents: z__default.default.record(z__default.default.string(), serializedAgentDefinitionSchema),
  workflows: z__default.default.record(z__default.default.string(), serializedWorkflowSchema),
  inputProcessors: z__default.default.array(serializedProcessorSchema),
  outputProcessors: z__default.default.array(serializedProcessorSchema),
  provider: z__default.default.string().optional(),
  modelId: z__default.default.string().optional(),
  modelVersion: z__default.default.string().optional(),
  modelList: z__default.default.array(modelConfigSchema).optional(),
  defaultOptions: z__default.default.record(z__default.default.string(), z__default.default.any()).optional(),
  defaultGenerateOptionsLegacy: z__default.default.record(z__default.default.string(), z__default.default.any()).optional(),
  defaultStreamOptionsLegacy: z__default.default.record(z__default.default.string(), z__default.default.any()).optional()
});
serializedAgentSchema.extend({
  id: z__default.default.string()
});
var providerSchema = z__default.default.object({
  id: z__default.default.string(),
  name: z__default.default.string(),
  label: z__default.default.string().optional(),
  description: z__default.default.string().optional()
});
var providersResponseSchema = z__default.default.object({
  providers: z__default.default.array(providerSchema)
});
var listAgentsResponseSchema = z__default.default.record(z__default.default.string(), serializedAgentSchema);
var listToolsResponseSchema = z__default.default.record(z__default.default.string(), serializedToolSchema);
var agentMemoryOptionSchema = z__default.default.object({
  thread: z__default.default.union([z__default.default.string(), z__default.default.object({ id: z__default.default.string() }).passthrough()]),
  resource: z__default.default.string(),
  options: z__default.default.record(z__default.default.string(), z__default.default.any()).optional(),
  readOnly: z__default.default.boolean().optional()
});
var toolChoiceSchema = z__default.default.union([
  z__default.default.enum(["auto", "none", "required"]),
  z__default.default.object({ type: z__default.default.literal("tool"), toolName: z__default.default.string() })
]);
var agentExecutionBodySchema = z__default.default.object({
  // REQUIRED
  messages: z__default.default.union([
    z__default.default.array(chunkMCYD5LW7_cjs.coreMessageSchema),
    // Array of messages
    z__default.default.string()
    // Single user message shorthand
  ]),
  // Message Configuration
  instructions: systemMessageSchema.optional(),
  system: systemMessageSchema.optional(),
  context: z__default.default.array(chunkMCYD5LW7_cjs.coreMessageSchema).optional(),
  // Memory & Persistence
  memory: agentMemoryOptionSchema.optional(),
  runId: z__default.default.string().optional(),
  savePerStep: z__default.default.boolean().optional(),
  // Request Context (handler-specific field - merged with server's requestContext)
  requestContext: z__default.default.record(z__default.default.string(), z__default.default.any()).optional(),
  // Execution Control
  maxSteps: z__default.default.number().optional(),
  stopWhen: z__default.default.any().optional(),
  // Model Configuration
  providerOptions: z__default.default.object({
    anthropic: z__default.default.record(z__default.default.string(), z__default.default.any()).optional(),
    google: z__default.default.record(z__default.default.string(), z__default.default.any()).optional(),
    openai: z__default.default.record(z__default.default.string(), z__default.default.any()).optional(),
    xai: z__default.default.record(z__default.default.string(), z__default.default.any()).optional()
  }).optional(),
  modelSettings: z__default.default.any().optional(),
  // Tool Configuration
  activeTools: z__default.default.array(z__default.default.string()).optional(),
  toolsets: z__default.default.record(z__default.default.string(), z__default.default.any()).optional(),
  clientTools: z__default.default.record(z__default.default.string(), z__default.default.any()).optional(),
  toolChoice: toolChoiceSchema.optional(),
  requireToolApproval: z__default.default.boolean().optional(),
  // Evaluation
  scorers: z__default.default.union([
    z__default.default.record(z__default.default.string(), z__default.default.any()),
    z__default.default.record(
      z__default.default.string(),
      z__default.default.object({
        scorer: z__default.default.string(),
        sampling: z__default.default.any().optional()
      })
    )
  ]).optional(),
  returnScorerData: z__default.default.boolean().optional(),
  // Observability
  tracingOptions: chunkMCYD5LW7_cjs.tracingOptionsSchema.optional(),
  // Structured Output
  output: z__default.default.any().optional(),
  // Zod schema, JSON schema, or structured output object
  structuredOutput: z__default.default.object({
    schema: z__default.default.object({}).passthrough(),
    model: z__default.default.union([z__default.default.string(), z__default.default.any()]).optional(),
    instructions: z__default.default.string().optional(),
    jsonPromptInjection: z__default.default.boolean().optional(),
    errorStrategy: z__default.default.enum(["strict", "warn", "fallback"]).optional(),
    fallbackValue: z__default.default.any().optional()
  }).optional()
}).passthrough();
var agentExecutionLegacyBodySchema = agentExecutionBodySchema.extend({
  resourceId: z__default.default.string().optional(),
  resourceid: z__default.default.string().optional(),
  // lowercase variant
  threadId: z__default.default.string().optional()
});
var executeToolDataBodySchema = z__default.default.object({
  data: z__default.default.custom((x) => x !== void 0, { message: "data is required" })
});
var executeToolBodySchema = executeToolDataBodySchema.extend({
  requestContext: z__default.default.record(z__default.default.string(), z__default.default.any()).optional()
});
var executeToolContextBodySchema = executeToolDataBodySchema.extend({
  requestContext: z__default.default.record(z__default.default.string(), z__default.default.any()).optional()
});
var voiceSpeakersResponseSchema = z__default.default.array(
  z__default.default.object({
    voiceId: z__default.default.string()
  }).passthrough()
  // Allow provider-specific fields like name, language, etc.
);
var toolCallActionBodySchema = z__default.default.object({
  runId: z__default.default.string(),
  requestContext: z__default.default.record(z__default.default.string(), z__default.default.any()).optional(),
  toolCallId: z__default.default.string(),
  format: z__default.default.string().optional()
});
var networkToolCallActionBodySchema = z__default.default.object({
  runId: z__default.default.string(),
  requestContext: z__default.default.record(z__default.default.string(), z__default.default.any()).optional(),
  format: z__default.default.string().optional()
});
var approveToolCallBodySchema = toolCallActionBodySchema;
var declineToolCallBodySchema = toolCallActionBodySchema;
var approveNetworkToolCallBodySchema = networkToolCallActionBodySchema;
var declineNetworkToolCallBodySchema = networkToolCallActionBodySchema;
var toolCallResponseSchema = z__default.default.object({
  fullStream: z__default.default.any()
  // ReadableStream
});
var updateAgentModelBodySchema = z__default.default.object({
  modelId: z__default.default.string(),
  provider: z__default.default.string()
});
var reorderAgentModelListBodySchema = z__default.default.object({
  reorderedModelIds: z__default.default.array(z__default.default.string())
});
var updateAgentModelInModelListBodySchema = z__default.default.object({
  model: z__default.default.object({
    modelId: z__default.default.string(),
    provider: z__default.default.string()
  }).optional(),
  maxRetries: z__default.default.number().optional(),
  enabled: z__default.default.boolean().optional()
});
var modelManagementResponseSchema = chunkMCYD5LW7_cjs.messageResponseSchema;
var generateSpeechBodySchema = z__default.default.object({
  text: z__default.default.string(),
  speakerId: z__default.default.string().optional()
});
var transcribeSpeechBodySchema = z__default.default.object({
  audio: z__default.default.any(),
  // Buffer
  options: z__default.default.record(z__default.default.string(), z__default.default.any()).optional()
});
var transcribeSpeechResponseSchema = z__default.default.object({
  text: z__default.default.string()
});
var getListenerResponseSchema = z__default.default.any();
var generateResponseSchema = z__default.default.any();
var streamResponseSchema = z__default.default.any();
var speakResponseSchema = z__default.default.any();
var executeToolResponseSchema = z__default.default.any();
var enhanceInstructionsBodySchema = z__default.default.object({
  instructions: z__default.default.string().describe("The current agent instructions to enhance"),
  comment: z__default.default.string().describe("User comment describing how to enhance the instructions")
});
var enhanceInstructionsResponseSchema = z__default.default.object({
  explanation: z__default.default.string().describe("Explanation of the changes made"),
  new_prompt: z__default.default.string().describe("The enhanced instructions")
});

exports.agentExecutionBodySchema = agentExecutionBodySchema;
exports.agentExecutionLegacyBodySchema = agentExecutionLegacyBodySchema;
exports.agentIdPathParams = agentIdPathParams;
exports.agentToolPathParams = agentToolPathParams;
exports.approveNetworkToolCallBodySchema = approveNetworkToolCallBodySchema;
exports.approveToolCallBodySchema = approveToolCallBodySchema;
exports.declineNetworkToolCallBodySchema = declineNetworkToolCallBodySchema;
exports.declineToolCallBodySchema = declineToolCallBodySchema;
exports.enhanceInstructionsBodySchema = enhanceInstructionsBodySchema;
exports.enhanceInstructionsResponseSchema = enhanceInstructionsResponseSchema;
exports.executeToolBodySchema = executeToolBodySchema;
exports.executeToolContextBodySchema = executeToolContextBodySchema;
exports.executeToolResponseSchema = executeToolResponseSchema;
exports.generateResponseSchema = generateResponseSchema;
exports.generateSpeechBodySchema = generateSpeechBodySchema;
exports.getListenerResponseSchema = getListenerResponseSchema;
exports.listAgentsResponseSchema = listAgentsResponseSchema;
exports.listToolsResponseSchema = listToolsResponseSchema;
exports.modelConfigIdPathParams = modelConfigIdPathParams;
exports.modelManagementResponseSchema = modelManagementResponseSchema;
exports.providersResponseSchema = providersResponseSchema;
exports.reorderAgentModelListBodySchema = reorderAgentModelListBodySchema;
exports.serializedAgentSchema = serializedAgentSchema;
exports.serializedToolSchema = serializedToolSchema;
exports.speakResponseSchema = speakResponseSchema;
exports.streamResponseSchema = streamResponseSchema;
exports.toolCallResponseSchema = toolCallResponseSchema;
exports.toolIdPathParams = toolIdPathParams;
exports.transcribeSpeechBodySchema = transcribeSpeechBodySchema;
exports.transcribeSpeechResponseSchema = transcribeSpeechResponseSchema;
exports.updateAgentModelBodySchema = updateAgentModelBodySchema;
exports.updateAgentModelInModelListBodySchema = updateAgentModelInModelListBodySchema;
exports.voiceSpeakersResponseSchema = voiceSpeakersResponseSchema;
