// @mastra/server/dist/chunk-MCYD5LW7.cjs
'use strict';

var z = require('zod');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var z__default = /*#__PURE__*/_interopDefault(z);

// src/server/schemas/common.ts
var runIdSchema = z__default.default.object({
  runId: z__default.default.string().describe("Unique identifier for the run")
});
var optionalRunIdSchema = z__default.default.object({
  runId: z__default.default.string().optional()
});
var paginationInfoSchema = z__default.default.object({
  total: z__default.default.number(),
  page: z__default.default.number(),
  perPage: z__default.default.union([z__default.default.number(), z__default.default.literal(false)]),
  hasMore: z__default.default.boolean()
});
var createPagePaginationSchema = (defaultPerPage) => {
  const baseSchema = {
    page: z__default.default.coerce.number().optional().default(0)
  };
  if (defaultPerPage !== void 0) {
    return z__default.default.object({
      ...baseSchema,
      perPage: z__default.default.coerce.number().optional().default(defaultPerPage)
    });
  } else {
    return z__default.default.object({
      ...baseSchema,
      perPage: z__default.default.coerce.number().optional()
    });
  }
};
var createCombinedPaginationSchema = () => {
  return z__default.default.object({
    page: z__default.default.coerce.number().optional(),
    perPage: z__default.default.coerce.number().optional(),
    /**
     * @deprecated Use page and perPage instead
     */
    offset: z__default.default.coerce.number().optional(),
    /**
     * @deprecated Use page and perPage instead
     */
    limit: z__default.default.coerce.number().optional()
  });
};
var tracingOptionsSchema = z__default.default.object({
  metadata: z__default.default.record(z__default.default.string(), z__default.default.unknown()).optional(),
  requestContextKeys: z__default.default.array(z__default.default.string()).optional(),
  traceId: z__default.default.string().optional(),
  parentSpanId: z__default.default.string().optional()
});
var coreMessageSchema = z__default.default.any();
var successResponseSchema = z__default.default.object({
  success: z__default.default.boolean()
});
var messageResponseSchema = z__default.default.object({
  message: z__default.default.string()
});
z__default.default.object({
  partial: z__default.default.string().optional()
});
var baseLogMessageSchema = z__default.default.object({
  level: z__default.default.enum(["debug", "info", "warn", "error", "silent"]),
  msg: z__default.default.string(),
  time: z__default.default.date(),
  context: z__default.default.record(z__default.default.string(), z__default.default.unknown()).optional(),
  runId: z__default.default.string().optional(),
  pid: z__default.default.number(),
  hostname: z__default.default.string(),
  name: z__default.default.string()
});

exports.baseLogMessageSchema = baseLogMessageSchema;
exports.coreMessageSchema = coreMessageSchema;
exports.createCombinedPaginationSchema = createCombinedPaginationSchema;
exports.createPagePaginationSchema = createPagePaginationSchema;
exports.messageResponseSchema = messageResponseSchema;
exports.optionalRunIdSchema = optionalRunIdSchema;
exports.paginationInfoSchema = paginationInfoSchema;
exports.runIdSchema = runIdSchema;
exports.successResponseSchema = successResponseSchema;
exports.tracingOptionsSchema = tracingOptionsSchema;
